### PR TITLE
fix(controller): Fixed workflow stuck with mutex lock

### LINKED ISSUE #4703: Workflows with mutexes get stuck
## Summary 

I'm trying to run a few workflows that are using a mutex, however if I run ~25 steps on the same mutex, the workflows get stuck and can't grab the mutex.

What happened/what you expected to happen?

The workflows get stucks and hangs indefinitely. Deleting the stuck workflows with the web-ui or kubectl does not release the mutex so new workflows can't run until the workflow-controller have been restarted.

## Diagnostics

What Kubernetes provider are you using? 

We are using Azure AKS with Kubernetes v1.18.10.

What version of Argo Workflows are you running? 

I have first encountered the issue with our argo v2.11.7 (bf3fec176cf6bdf3e23b2cb73ec7d4e3d051ca40) with a cluster install.
Then i tested this in a different k8s cluster (still on azure) with argo v2.11.6, v2.11.7 and v2.11.8 in a namespaced environment.
The logs in this issue will be from my test argo workow v2.11.8 (310e099f82520030246a7c9d66f3efaadac9ade2).

The issue is only appearing with this specific workflowtemplate architecture. The result is mostly random.

WorkflowTemplate used to reproduce the issue:
```yaml
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: test
spec:
  entrypoint: entrypoint
  onExit: on-exit
  templates:
  - name: entrypoint
    steps:
    - - name: generate-jobs
        template: generate-jobs
    - - arguments:
          parameters:
          - name: mutex_name
            value: '{{item}}'
        name: test
        template: main-function
        withParam: '{{steps.generate-jobs.outputs.result}}'

  - name: generate-jobs
    script:
      command: [ cat ]
      image: alpine:3.6
      source: |
        ["config", "config", "config", "config", "config"]

  - name: main-function
    inputs:
      parameters:
      - name: mutex_name
    synchronization:
      mutex:
        name: '{{inputs.parameters.mutex_name}}'
    steps:
    - - name: check
        templateRef:
          name: test
          template: main-command

  - name: main-command
    container:
      command: [ sh ]
      args: [ "-c", "sleep 5" ]
      image: alpine:3.6

  - name: on-exit
    script:
      command:
      - cat
      image: alpine:3.6
      source: |
        echo fixme
```

Shell command to trigger the issue:
```
for i in $(seq 5); do argo submit --from workflowtemplate/test ; done
```

I take as an example the case where I launch 5 workflows from the workflow template.
One workflow has succeeded, but all others are stuck.
(I'm using github gists to keep the issue readable)

The workflow that succeeded: [GIST](https://gist.githubusercontent.com/Fumesover/0360ea227560f4212e1a22ac40fda4b4/raw/11a9a1ffc9f39345a6170ee14eb4a47d6ca7eebc/test-d44k6.yaml)
A stuck workflow: [GIST](https://gist.githubusercontent.com/Fumesover/0360ea227560f4212e1a22ac40fda4b4/raw/11a9a1ffc9f39345a6170ee14eb4a47d6ca7eebc/test-hkkhv.yaml)
workflow-controller logs: [GIST](https://gist.githubusercontent.com/Fumesover/0360ea227560f4212e1a22ac40fda4b4/raw/11a9a1ffc9f39345a6170ee14eb4a47d6ca7eebc/workflow-controller-79d879b95d-q6fqk.logs)

Apparently, the mutex is still held by one of the "argo nodes" (and not by a pod). But I can't figure out what kind of object it is in Kubernetes, so I don't know where to look for the cause of this behavior.

I'm on the argoproj slack (`@Albin Parou`) if you have any questions regarding this issue. :slightly_smiling_face: 

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by sarabala1979**
Signed-off-by: Saravanan Balasubramanian <sarabala1979@gmail.com>

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Fixes #4703

### CODE REVIEW

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -2612,8 +2612,10 @@ func (ms *MutexStatus) LockAcquired(holderKey, lockKey string, currentHolders []
 }
 
 func (ms *MutexStatus) LockReleased(holderKey, lockKey string) bool {
-	i, _ := ms.GetHolding(lockKey)
-	if i >= 0 {
+	i, holder := ms.GetHolding(lockKey)
+	items := strings.Split(holderKey, "/")
+	holdingName := items[len(items)-1]
+	if i >= 0 && holder.Holder == holdingName {
```
Comment (sarabala1979) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:2618]: This check will make sure who is requesting release

File: workflow/controller/steps.go [hunk 2]
```diff
@@ -93,7 +93,11 @@ func (woc *wfOperationCtx) executeSteps(nodeName string, tmplCtx *templateresolu
 
 		sgNode := woc.executeStepGroup(stepGroup.Steps, sgNodeName, &stepsCtx)
 
-		if !sgNode.Fulfilled() {
+		if sgNode.Fulfilled() {
```
Comment (sarabala1979) [inline workflow/controller/steps.go:96]: If the step template has mutex lock and will not invoke execute template to release the lock if child nodes are completed.  It. has to be handled here
Eg:
```yaml
- name: main-function
    synchronization:
      mutex:
        name:  config
    steps:
    - - name: check
        template: main-command
    - - name: check1
        template: main-command


```

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 3]
```diff
@@ -2612,8 +2612,10 @@ func (ms *MutexStatus) LockAcquired(holderKey, lockKey string, currentHolders []
 }
 
 func (ms *MutexStatus) LockReleased(holderKey, lockKey string) bool {
-	i, _ := ms.GetHolding(lockKey)
-	if i >= 0 {
+	i, holder := ms.GetHolding(lockKey)
+	items := strings.Split(holderKey, "/")
+	holdingName := items[len(items)-1]
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:2617]: What if `len(items) == 0`?

**Review by simster7 — CHANGES_REQUESTED** [review]

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 3 above)
Comment (sarabala1979) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:2617]: Holderkey will have two default elements in synchronization scenario. Namespace/wf name. I can add a  check

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 4]
```diff
@@ -2511,6 +2511,9 @@ func (ss *SemaphoreStatus) LockAcquired(holderKey, lockKey string, currentHolder
 	i, semaphoreHolding := ss.GetHolding(lockKey)
 	items := strings.Split(holderKey, "/")
 	holdingName := items[len(items)-1]
+	if len(items) == 0 {
+		return false
+	}
```
Comment (simster7) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:2516]: I think you mean to put this above live 2513

**Review by simster7 — CHANGES_REQUESTED** [review]

**Review by simster7 — APPROVED** [review]

**Pull request merged**